opds2_import.py

```diff
@@ -0,0 +1,830 @@
+            metadata_client,
+            content_modifier,
+            map_from_collection,
+            mirrors,
+        )
+
+        self._logger = logging.getLogger(__name__)
+
+    def _extract_subjects(self, subject):
+        """Extract a list of SubjectData objects from the webpub-manifest-parser's subject.
+
+        :param subject: Parsed subject object
+        :type subject: Union[
+            str,
+            List[Union[str, core_ast.Subject]],
+            core_ast.Subject
+        ]
+
+        :return: List of subjects metadata
+        :rtype: List[SubjectMetadata]
+        """
+        self._logger.debug(
+            u"Started extracting subject metadata from {0}".format(encode(subject))
+        )
+
+        subject_metadata_list = []
+
+        if is_string(subject):
+            subject_metadata = SubjectData(
+                type=None, identifier=None, name=subject, weight=1
+            )
+            subject_metadata_list.append(subject_metadata)
+        elif isinstance(subject, core_ast.Subject):
+            scheme = subject.scheme
+
+            subject_type = Subject.by_uri.get(scheme)
+            if not subject_type:
+                # We can't represent this subject because we don't
+                # know its scheme. Just treat it as a tag.
+                subject_type = Subject.TAG
+
+            subject_metadata = SubjectData(
+                type=subject_type, identifier=subject.code, name=subject.name, weight=1
+            )
+
+            subject_metadata_list.append(subject_metadata)
+        elif isinstance(subject, list):
+            for inner_subject in subject:
+                inner_subject_metadata_list = self._extract_subjects(inner_subject)
+
+                subject_metadata_list.extend(inner_subject_metadata_list)
+
+        self._logger.debug(
+            u"Finished extracting contributor metadata from {0}: {1}".format(
+                encode(subject), encode(subject_metadata_list)
+            )
+        )
+
+        return subject_metadata_list
+
+    def _extract_contributors(self, contributor):
+        """Extract a list of ContributorData objects from the webpub-manifest-parser's contributor.
+
+        :param contributor: Parsed contributor object
+        :type contributor: Union[
+            str,
+            List[Union[str, core_ast.Contributor]],
+            core_ast.Contributor
+        ]
+
+        :return: List of contributors metadata
+        :rtype: List[ContributorData]
+        """
+        self._logger.debug(
+            u"Started extracting contributor metadata from {0}".format(
+                encode(contributor)
+            )
+        )
+
+        contributor_metadata_list = []
+
+        if is_string(contributor):
+            contributor_metadata = ContributorData(
+                sort_name=None,
+                display_name=contributor,
+                family_name=None,
+                wikipedia_name=None,
+                roles=None,
```

Comment:
I assume this is the situation where the OPDS feed lists a single string as the contributor. When you call _extract_contributors I mention that you should pass in a default role; this is where the default role should be used.

Reply:
Updated the code